### DISCUSSION TITLE
Cannot cluster two nodes

### DISCUSSION THREAD

**Discussion by jlalanda** [rabbitmq/rabbitmq-server#3009] [Other]
Hello,

I have posted my problem on the google groups and didn´t got any response.

So this is my problem:

I cannot create a cluster of two windows machines running the same rabbitmq Version and erlang version. (RabbitMQ 3.8.14 and Erlang 23.3.1)

-Both machines ping each other by ip and by its hostname. And for reassuring that i configured the ip and FQDN on the hosts file
-cookie is exactly the same between the folders of the user running the windows service and the windows folder
-All of the ports are opened between both machines (i have done the telnet client test for all ports)
-Both nodes run rabbitmq successfully 
-Also run the following commandos stop_app, reset and start_app on the other onde just to guarantee that is all ok.

When i try to join clusters this is the error im getting:

Clustering node rabbit@Machine1 with rabbit@Machine2
Error:
{:badarg, [{:rpc, :rpcify_exception, 2, [file: 'rpc.erl', line: 467]}, {:rpc, :call, 5, [file: 'rpc.erl', line: 410]}, {:lists, :foldl, 3, [file: 'lists.erl', line: 1267]}, {:rabbit_mnesia, :discover_cluster, 1, [file: 'src/rabbit_mnesia.erl', line: 795]}, {:rabbit_mnesia, :join_cluster, 2, [file: 'src/rabbit_mnesia.erl', line: 227]}]}

This are the commands i am doing:
-reset
-stop_app
-join_cluster rabbit@Machine2

This is the group i posted my problem:
https://groups.google.com/g/rabbitmq-users/c/tQBttqsQKz4/m/AbQBqAvcAwAJ

Can you please check what is the problem?

**Comment by michaelklishin** [comment]
I will convert this issue to a GitHub discussion. Currently GitHub will [automatically close and lock the issue](https://github.com/github/feedback/discussions/3197) even though your question will be transferred and responded to elsewhere. This is to let you know that we do not intend to ignore this but this is how the current GitHub conversion mechanism makes it seem for the users :(

**Reply by jlalanda** [reply]
Ok. I just hope somebody will see this because i need to create a cluster for this product to work on my project

**Comment by michaelklishin** [comment]
I'm afraid there isn't enough information in this stack trace to tell what prevents nodes from clustering. My best guess is that the "node" argument is not an atom.

You haven't provided a list of steps to reproduce. How exactly are you clustering these nodes? Is it using CLI tools or a [peer discovery mechanism](https://www.rabbitmq.com/cluster-formation.html)? If it's the latter, what is in the config file? You can be using the wrong data type for node names.

What is in the logs of both nodes?

**Reply by jlalanda** [reply]
I dont have any logs for the command im doing (join_cluster)

**Reply by michaelklishin** [reply]
What is the exact command you use? We do not guess in this community. Please help others help you.

**Comment by michaelklishin** [comment]
For nodes to be able to join each other, they must satisfy a number of [requirements](https://www.rabbitmq.com/clustering.html#cluster-formation-requirements):

 * Be able to resolve each others hostname.
 * Use the same shared secret (Erlang cookie). If this is the failing step, target node logs will contain a message about connections from a node that fails authentication
 * Have port 25672 open on both sides
 * The joining node must be reset first or be a blank starting node

`ping` is not particularly useful when assessing network connectivity, see [Troubleshooting Networking](https://www.rabbitmq.com/troubleshooting-networking.html).

I recommend using the [classic config peer discovery](https://www.rabbitmq.com/cluster-formation.html) option instead of CLI tools.
The joining node must be reset between attempts, as the guide explains: otherwise it will not perform peer discovery. There
will be a lot of relevant information logged [at debug level](https://www.rabbitmq.com/logging.html#debug-logging).

Without a detailed set of steps you use, we cannot suggest much more. There is no evidence of a RabbitMQ or runtime bug of any kind.

**Reply by jlalanda** [reply]
With the initial text i indicated the following:

-Both Machines resolve each other hostname using nslookup tool
-The have all the ports opened (the command telnet is working with all pports)
-The joining is being made with both machine being reset before using the reset command
-The cookie was copied from one machine to the other and placed on the user folder and also on the windows folder.

**Reply by michaelklishin** [reply]
Great. How are you joining them exactly? Have you considered using peer discovery instead of the manual process?

**Comment by michaelklishin** [comment]
You can add `--print-stacktrace` to any command so that it prints the entire stack trace on the CLI side. The one posted doesn't reveal any arguments passed to the function that throws, unfortunately. It's not a given that `--print-stacktrace` will print more but it is worth trying.

If instead of manual clustering you'd use classic config peer discovery, nodes will log any errors, and that might contain additional context. Note that both nodes will [use a startup delay](https://www.rabbitmq.com/cluster-formation.html#initial-formation-race-condition) when starting as blank nodes with peer discovery configured, by design. This should not surprise you.

**Reply by jlalanda** [reply]
Hello,

I am back again. I have done the --print-stacktrace again and the error is the same

[root@Machine02]# sudo rabbitmqctl join_cluster rabbit@Machine1 --print-stacktrace
Clustering node rabbit@Machine02 with rabbit@Machine01
Error:
{:badarg, [{:rpc, :rpcify_exception, 2, [file: 'rpc.erl', line: 467]}, {:rpc, :call, 5, [file: 'rpc.erl', line: 410]}, {:l
ists, :foldl, 3, [file: 'lists.erl', line: 1267]}, {:rabbit_mnesia, :discover_cluster, 1, [file: 'src/rabbit_mnesia.erl', 
line: 795]}, {:rabbit_mnesia, :join_cluster, 2, [file: 'src/rabbit_mnesia.erl', line: 227]}]}

What information do you need more? So we can discover whats wrong.

Just one note i changed my original setup from windows machines to linux machines (redhat) in hope that the problem was specific to windows....

**Reply by jlalanda** [reply]
when i try to run the command 

rabbitmq-diagnostics status -n rabbit@Machine01 -l

From Machine 02 i indeed get the following message:

DIAGNOSTICS
===========

attempted to contact: ['rabbit@Machine01']

rabbit@Machine01:
  * connected to epmd (port 4369) on Machine01
  * epmd reports node 'rabbit' uses port 25672 for inter-node and CLI tool traffic 
  * TCP connection succeeded but Erlang distribution failed 
  * suggestion: check if the Erlang cookie identical for all server nodes and CLI tools
  * suggestion: check if all server nodes and CLI tools use consistent hostnames when addressing each other
  * suggestion: check if inter-node connections may be configured to use TLS. If so, all nodes and CLI tools must do that
   * suggestion: see the CLI, clustering and networking guides on https://rabbitmq.com/documentation.html to learn more


Current node details:
 * node name: 'rabbitmqcli-726-rabbit@Machine01'
 * effective user's home directory: /var/lib/rabbitmq
 * Erlang cookie hash: +jSCZqNnPaAsNz2keFQWYQ==
 
 I have compared the Erlang Cookie hash and checked out that is equal on both machines, so that is not the problem.
 
 Also both machines have the following doors opened on firewall:
 4369/tcp 5671/tcp 5672/tcp 25672/tcp 35672-35682/tcp 15672/tcp

Where can i get more information to check where is the problem?

**Reply by michaelklishin** [reply]
If you are certain that the shared secret (cookie file) content is identical and TLS is not used, this leaves mismatching hostnames as the third most likely candidate that the error message lists:

> suggestion: check if all server nodes and CLI tools use consistent hostnames when addressing each other

When a node starts, it will logs its node name, data directory, config file locations (and whether the files were found) and so on. You can use this to see what names the nodes use. For `rabbit@machine1` and `rabbit@machine2` to be able to cluster,
node `rabbit@machine1` must identify itself as `rabbit@machine1` (and not, say, `rabbit@localhost`) and so should its cluster peers. This will be the case for all nodes. If node `rabbit@A` connects to `rabbit@B` and the "B node" identifies itself as `rabbit@C`, clustering will fail with

> TCP connection succeeded but Erlang distribution failed

which is the most precise kind of error the runtime gives us. The target node (the one being joined) will log an warning
about a "connection from a disallowed node" or something similar.

Consider inspecting and sharing node logs. Note that resetting a node will delete all of its data and in some cases (I believe on Windows specifically) this includes logs. Guessing as to what is going on can take a lot of time.
